lib/pka_dev.c

```diff
@@ -1928,25 +1939,41 @@ int __pka_dev_open_ring(uint32_t ring_id)
         return -EPERM;
 
     ring = pka_dev_get_ring(ring_id);
+
     if (!ring || !ring->shim)
         return -ENXIO;
 
     shim = ring->shim;
 
+    mutex_lock(&ring->mutex);

```

Comment:
I believe we can postpone acquiring the lock to line 1957, after we check the shim status and before we check the ring status. What do you think?

Reply:
It will work, yes. Since we are also updating the `shim->status` while holding `ring->mutex` lock (line 1980 and 2141), I was thinking its better to check the `shim->status` while holding this lock so that if we add some more shim->status checks in the future we don't have to move the code for acquiring the lock.
Or we can separate the `shim->status` and other shim updates `(shim->busy_ring_num`) from ring updates to avoid this confusion. Right now these updates are somewhat intertwined during open and close ring operations. After separation we can encapsulate lock and unlock around ring related checks and updates only.
Let me know what do you think.

Reply:
Yep you are absolutely right! I missed it. In that case let's keep it as is.